catkin_ws/src/state_estimation/src/sensors.py

```diff
@@ -19,111 +18,121 @@
+    def depth_cb(self, msg):
+        self.z = msg.data + self.z_pos_mount_offset
+
+    def get_reading(self):
+        return {
+            "z": self.z
+        }
 
+
+# IMU class inheriting from Sensor
 class IMU(Sensor):
     def __init__(self):
         super().__init__("IMU")
-
+        
         q_imunominal_imu_w = rospy.get_param("q_imunominal_imu_w")
         q_imunominal_imu_x = rospy.get_param("q_imunominal_imu_x")
         q_imunominal_imu_y = rospy.get_param("q_imunominal_imu_y")
         q_imunominal_imu_z = rospy.get_param("q_imunominal_imu_z")
 
         self.q_imunominal_imu = np.quaternion(q_imunominal_imu_w, q_imunominal_imu_x, q_imunominal_imu_y, q_imunominal_imu_z)
         self.q_imu_auv = self.q_imunominal_imu.conjugate() * Q_IMUNOMINAL_AUV
-        self.q_nwu_auv = None
-        
+        self.q_nwu_auv = np.quaternion(1, 0, 0, 0)
+
+        self.angular_velocity = np.array([0,0,0])
+
+        rospy.Subscriber("/sensors/imu/status", Int32, self.sensor_status_cb)
         rospy.Subscriber("/sensors/imu/angular_velocity", SbgImuData, self.ang_vel_cb)
         rospy.Subscriber("/sensors/imu/quaternion", SbgEkfQuat, self.quat_cb)
-        
+
     def ang_vel_cb(self, msg):
         # angular velocity vector relative to imu frame 
         ang_vel_imu = np.array([msg.gyro.x, msg.gyro.y, msg.gyro.z])
         # anuglar velocity vector relative to AUV frame 
         temp_angular_velocity = self.q_imu_auv * np.quaternion(0, ang_vel_imu[0], ang_vel_imu[1], ang_vel_imu[2]) * self.q_imu_auv.conjugate()
         self.angular_velocity = np.array([temp_angular_velocity.x, temp_angular_velocity.y, temp_angular_velocity.z])
-        self.updateLastState()
-
+    
     def quat_cb(self, msg):
         q_ned_imu = np.quaternion(msg.quaternion.w, msg.quaternion.x, msg.quaternion.y, msg.quaternion.z)
-        # print(q_ned_imu)
         q_nwu_imu = Q_NWU_NED * q_ned_imu
         self.q_nwu_auv = q_nwu_imu * self.q_imu_auv
-        self.updateLastState()
 
-    def hasValidData(self):
-        return self.q_nwu_auv is not None and self.angular_velocity is not None
+    def get_reading(self):
+        return {
+            "quaternion": self.q_nwu_auv, 
+            "angular_velocity": self.angular_velocity
+        }
 
+
+# IMUFrontCamera class inheriting from Sensor
+class IMUFrontCamera(Sensor):
+    def __init__(self):
+        super().__init__("IMU Front Camera")
+
+        self.q_nwu_auv = np.quaternion(1, 0, 0, 0)
+        self.angular_velocity = np.array([0,0,0])
+
+        rospy.Subscriber("/sensors/imu_front_camera/status", Int32, self.sensor_status_cb)
+        rospy.Subscriber("/zed/zed_node/imu/data", Imu, self.front_camera_imu_cb)
+
+    def front_camera_imu_cb(self, msg):
+        self.q_nwu_auv = np.quaternion(msg.orientation.w, msg.orientation.x, msg.orientation.y, msg.orientation.z)
+        self.angular_velocity = np.array([msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z]) 
+
```

Comment:
I have no idea what the transformations are supposed to fix the reference frame, so I just set the variable to the new reading. We also need to check this in person